### PR TITLE
GSoC 2019: Patient search criteria added in openmrs core

### PR DESCRIPTION
**Pull request by Reyano132**
The functionality of patient search criteria module added in openmrs core. This is part of my GSoC 2019 Project. https://wiki.openmrs.org/display/projects/Patient+Search+Criteria

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 1]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+		return patients;
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query, String gender, Integer start, Integer length, Boolean includeVoided)
+	        throws DAOException {
+		List<Patient> patients = findPatients(query, includeVoided, start, length);
+		List<Patient> result=new LinkedList<>();
+		for (Patient p : patients) {
+			if (p.getGender().equals(gender)) {
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: This filtering can be done using java8 stream filtering. Example

`findPatients(query, includeVoided, start, length).stream().filter(patient -> patient.getGender().equals(gender)).collect(Collectors.toList());  `

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 2]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+			tmpLength = maxLength;
+		}
+		query = LuceneQuery.escapeQuery(query);
+
+		List<Patient> patients = new LinkedList<>();
+
+		String minChars = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_MIN_SEARCH_CHARACTERS);
+
+		if (minChars == null || !StringUtils.isNumeric(minChars)) {
+			minChars = "" + OpenmrsConstants.GLOBAL_PROPERTY_DEFAULT_MIN_SEARCH_CHARACTERS;
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: any reason to append "" in the beginning ?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 3]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+		for (Patient p : patients) {
+			if (p.getGender().equals(gender)) {
+				result.add(p);
+			}
+		}
+		return result;
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query, Date from, Date to, Integer start, Integer length,
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: Since this is a long method It would be better to add some comments in this method body by explaining whats happening .

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 4]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+			tmpLength -= personNames.getList().size();
+			tmpStart = 0;
+		} 
+		
+		return patients;
+
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query,Date birthdate, Integer start, Integer length,
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: this methods body is quite similar to the above getPatients method body and have lots of duplicate code.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 5]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query, String gender, Date from, Date to, Integer start, Integer length,
+	        Boolean includeVoided) throws DAOException {
+		List<Patient> patients = getPatients(query, from, to, start, length, includeVoided);
+		List<Patient> result= new LinkedList<>();
+		for (Patient p : patients) {
+			if (p.getGender().equals(gender)) {
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: please use java8 stream filtering

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 6]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+		return result;
+	}
+	
+	@Override
+	public List<Patient> getPatients(String query, String gender, Date birthdate, Integer start, Integer length,
+	        Boolean includeVoided) throws DAOException {
+		List<Patient> patients = getPatients(query, birthdate, start, length, includeVoided);
+		List<Patient> result=new LinkedList<>();
+		for (Patient p : patients) {
+			if (p.getGender().equals(gender)) {
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: please use java8 stream filtering

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 7]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+		List<Patient> patients = new LinkedList<>();
+		List<Patient> result=new LinkedList<>();
+		
+		PatientLuceneQuery personLuceneQuery = new PatientLuceneQuery(sessionFactory);
+		LuceneQuery<Person> birthdateQuery = personLuceneQuery.getPatinetWithBirthdate(birthdate, includeVoided);
+		long namesSize = birthdateQuery.resultSize();
+		if (namesSize > tmpStart) {
+			ListPart<Object[]> patientNames = birthdateQuery.listPartProjection(tmpStart, tmpLength, "personId");
+			patientNames.getList().forEach(patientName -> patients.add(getPatient((Integer) patientName[0])));
+			for(Patient p:patients) {
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: please use java8 stream filtering

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 8]
```diff
@@ -216,6 +218,71 @@ public Patient getPatient(Integer patientId) throws APIException {
 		return dao.getPatient(patientId);
 	}
 	
+	/**
+	 * @see org.openmrs.api.PatientService#getPatients(String, String, List, boolean, String, Integer, Integer, Date)
+	 */
+	
+	@Override
+	public List<Patient> getPatients(String name, String identifier, List<PatientIdentifierType> identifierTypes,
```
Comment (dilanthas) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: I think this method should be break in to several other methods: With all these if and else conditions its really hard to read whats going on.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java (hunk 2 above)
Comment (Reyano132) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: For casting int to string, although I removed that part

**Comment by dkayiwa** [comment]
@Reyano132 did you see the above review comments?

File: api/src/main/java/org/openmrs/api/db/PatientDAO.java [hunk 9]
```diff
@@ -297,4 +298,99 @@
...
+	        throws DAOException;
+	
+	/**
+	 * @param birthdate: birthdate of patient
+	 * @return list of patients who follow the query regex and having birthdate as required.
+	 * @throws DAOException
+	 */
+	public List<Patient> getPatients(String query, Date birthdate, Integer start, Integer length, Boolean includeVoided)
+	        throws DAOException;
+
```
[the remark this answers is not in the document]
Comment (Reyano132) [reply api/src/main/java/org/openmrs/api/db/PatientDAO.java]: Hey @dilanthas, I changed the names of methods as per inputs

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 10]
```diff
@@ -813,6 +813,321 @@ public Long getCountOfPatients(String query, boolean includeVoided) {
...
+		
+		long namesSize = birthdateQuery.resultSize();
+		if (namesSize > tmpStart) {
+			ListPart<Object[]> patientBirthdate = birthdateQuery.listPartProjection(tmpStart, tmpLength, "personId");
+			patientBirthdate.getList().forEach(patient -> patients.add(getPatient((Integer) patient[0])));
+		} 
+
+		return patients;
+	}
+
```
[the remark this answers is not in the document]
Comment (Reyano132) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: Also, I removed code duplication.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java (hunk 1 above)
Comment (Reyano132) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: thanks for the help, I use stream filter as you mention.

**Comment by Reyano132** [comment]
yes, @dkayiwa, I'm working on it

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java (hunk 3 above)
Comment (Reyano132) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java]: I improved the method

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 8 above)
Comment (Reyano132) [reply api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: I divide this method into two methods, now it becomes easy to understand

**Comment by Reyano132** [comment]
@dkayiwa  sir, can you please merge this PR. It's ready and I have to use this feature in core apps module

**Comment by dkayiwa** [comment]
@Reyano132  have your mentors reviewed and approved it?

**Comment by kaweesi** [comment]
hi @Reyano132, this looks impressive at first glance, i need to test it as soon as i can. I have some suggestions but in a meantime, please move all these other changes except for ones about the SEARCH_INDEX_VERSION version, Person.java and Patient.java into our search criteria module.

i have created https://github.com/openmrs/openmrs-module-patientsearchcriteria please after migrating these changes into your module, point git to the new and merge it into your local branch, open up one pull request from https://github.com/Reyano132/openmrs-module-patientsearchcriteria pointing to this new repository.

I am also looking forward to seeing your functional UI changes.
Kind regards

**Comment by HerbertYiga** [comment]
@Reyano132   Have you seen the above comments for @k-joseph ?

**Comment by Reyano132** [comment]
@kaweesi sir, As you said I changed this PR. Also, I create new PR for patientsearchcriteria, here : https://github.com/openmrs/openmrs-module-patientsearchcriteria/pull/1

**Comment by Reyano132** [comment]
@dkayiwa and @kaweesi  sir, I solved the problem of coverage/coveralls issue by adding the unit test case related to person index. Can you please merge this PR?

**Comment by Reyano132** [comment]
@kaweesi and @dkayiwa  sir can please merge this PR?  I'm running out of time

File: api/src/main/java/org/openmrs/util/OpenmrsConstants.java [hunk 11]
```diff
@@ -599,9 +599,9 @@ public static String getOpenmrsProperty(String property) {
 	/**
 	 * Indicates the version of the search index. The index will be rebuilt, if the version changes.
 	 * 
-	 * @since 1.11
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: Why did you change this?

Comment (Reyano132) [reply api/src/main/java/org/openmrs/util/OpenmrsConstants.java]: when we build the openmrs core, it will generate a person's Lucene index

**Comment by dkayiwa** [comment]
@Reyano132 can you bring this again to the attention of your primary mentor?
